### PR TITLE
Support `#to_s` on frozen Group::Passive

### PR DESCRIPTION
**Pull request by dgollahon**
On [mutant](https://github.com/mbj/mutant) we aggressively freeze objects and work with immutable structures. In trying to [re-introduce support](https://github.com/mbj/mutant/pull/1166) for (immutably) mutating regular expressions I ran into an issue with the 2.x series of `regexp_parser` that prevented calling `#to_s` on frozen passive groups. The change I am having trouble with was introduced [here](https://github.com/ammar/regexp_parser/commit/f14a36bff6fef3ef174b63c744c556469a42a589#diff-5bfe3942df5a59a97d7e5c7d3e918a1c704f964685e719c178af22401da46c23R25).

I am proposing a simple fix here that should hopefully not impact anything else but does allow calling `#to_s` on a frozen passive group.

@jaynetics could you let me know what you think about this issue/approach? Thanks again with your help on the last two issues.

### CODE REVIEW

**Comment by dgollahon** [comment]
Actually my test does not assert what I think it asserts--let me double check something.

File: spec/parser/groups_spec.rb [hunk 1]
```diff
@@ -105,4 +105,15 @@
...
     expect(root[2][1][0][1][0].number_at_level).to eq 1
   end
+
+  specify('supports #to_s when frozen') do
+    group_token, literal_token = Regexp::Lexer.lex(/(?:foo)/)
+
+    passive_group =
+      Group::Passive.new(group_token).tap do |expression|
+        expression.expressions = [Literal.new(literal_token)]
+      end.freeze
```
Comment (dgollahon) [inline spec/parser/groups_spec.rb:115]: So I learned something interesting when trying to set up this regression test: the issue only seems to occur when the group is manually constructed from tokens like this (this has to happen in `mutant` because of how we generate new ASTs and map them back into concrete `Regexp::...` objects and then use those objects to unparse the AST). If you use `RP.parse(/(?:foo)/)` and then call `#freeze` and then `#to_s` that actually works on master. So it seems like something normally initializes the `@implicit` ivar.

It is pretty difficult to manually construct the right regexp objects so this has been tricky for me to debug and figure out how to reproduce the issue.

Comment (jaynetics) [reply spec/parser/groups_spec.rb:115]: @dgollahon

I was intrigued by this because it seemed to suggest we might inadvertently be calling `#to_s` on expressions while parsing, as that would explain `@implicit` being initialized.

I could not reproduce what you said on master, though. For me, the error occurs with a frozen parse result, too.

spec:

```ruby
  specify('supports #to_s when frozen') do
    group = Regexp::Parser.parse(/(?:foo)/)[0]
    group.freeze
    expect(group.to_s).to eq '(?:foo)'
  end
```

output:

```
Failures:

  1) Group parsing supports #to_s when frozen
     Failure/Error: @implicit ||= false
     
     FrozenError:
       can't modify frozen Regexp::Expression::Group::Passive: ...
```

before you go and change the spec:

I've been thinking about how to have a bit more protection from such problems in the future. My idea is to create a regexp with all or at least a lot of the existing features, parse it (`parse(pattern, :any)`), recursively freeze all expressions, and see if it can be stringified and matches the input. `to_s_spec.rb` would be the place for that. This could actually be a bit of fun, do you want to do it? Otherwise, no problem, I'll merge and release this as is and add work on the spec when I get to it.

Comment (dgollahon) [reply spec/parser/groups_spec.rb]: 🤦 ok I think I see what happened here. I think I was freezing the result of of `Regexp::Expression::Root` which is what `RP.parse` returns instead of the first child and somehow never got to the right object to freeze. I tried it several different ways and I think must have just gotten mixed up or was temporarily on the wrong branch or something. Thanks for checking my work here!

I have done a first pass at what I think you have suggested: I updated all the existing assertions in the `to_s_spec` to use deeply frozen parsed objects. To do this easily I added the `ice_nine` gem as a development dependency. If that's a problem / not what you had in mind, let me know! I think that this approach is much better than what I originally had.

**Comment by dgollahon** [comment]
@jaynetics This is ready for another pass 😄 

I think the `regexp_parser` tests seem to cover a lot of the important cases but if you are interested in increasing robustness I would be happy to help out:

As one thought `regexp_parser` might benefit from corpus tests as suggested [here](https://github.com/ammar/regexp_parser/pull/12#issuecomment-219311873). We could add some machinery to see if `regexp_parser` can parse all the regexes in rubyspec and/or other common projects (rubocop?) and additional assert that they can be roundtripped on most or all cases.

Also, if you have any interest in having mutant as an (optional?) build step I would be happy to get an OSS license and set that up for this project. I find it can be very helpful for ensuring a thorough test suite, finding bugs, and simplifying code.

Regular code coverage tools might be another reasonable lever to pull to expand the test suite as well (mainly if `mutant` is not of interested since `mutant` coverage is generally a superset of regular code coverage).

If none of those suggestions appeal, that's totally fine by me too.

**Comment by mbj** [comment]
Also happy to assist in setting up mutant here. Given mutant will depend on `regexp_parser` again in the future.

I wounder if we need to initially disable regexp mutations wen mutant is in use here, as I doubt that `regexp_parser` is ready to use zombification (a term I invented to describe mutants self test system that recursively in-memory vendors dependencies).

Zombification has some pre-conditions that are not easily retrofitted to dependencies.

**Comment by jaynetics** [comment]
@dgollahon

Thanks again, that looks great! I've just released it as v2.0.2.

@dgollahon @mbj

I'm a big fan of mutant. I've used it quite a bit and learned lots about the finer nuances of Ruby.

Whenever you have any questions about your new integration, or if you want me to review it when it's ready, just drop me a note.

That being said, I'm a bit skeptical about using mutant on regexp_parser.

This is an old code base, the code is quite procedural, stuff resembling OOP only exists at the consumable surface layer, even in that layer spec files hardly correspond to any classes, and I probably won't have the time to substantially refactor it.

What is more, the scanner written in Ragel is the heart of the gem. It does get transpiled to Ruby, but that is [3000 lines of spaghetti Ruby jumping through machine state codes](https://gist.github.com/jaynetics/b667cabdce5af71f8e60683eef41ddde).

That might be an interesting test case for mutant (e.g.: will it take super long to run?), or even uncover imperfections in Ragel, but as most of the code isn't human readable, I'm not sure I'd want to fight mutants in there.

Corpus tests sound more feasible to me. Quickly skipping over ruby-spec gives me the impression that the regexp_parser specs already cover many more edge cases than ruby-spec, so I'm not sure that's worth the effort. Onigmo has [a lot of tests, conveniently written in Ruby](https://github.com/k-takata/Onigmo/blob/master/test.rb). rubocop also sounds good to me. capybara is another major dependent, but their regexps [are fairly tame](https://github.com/teamcapybara/capybara/blob/master/spec/regexp_dissassembler_spec.rb).

It would be awesome if you have any time to spare for this, or relevant reusable code to share!

**Comment by dgollahon** [comment]
@jaynetics Thanks for the release!

I think there might be some value-add from running `mutant` on the parts that can have it run but noted that it might be a bit more complicated. I think there is probably extra config needed here but I suspect something would work decently well. I am happy to do any legwork there to experiment if you're interested in it but noted that it doesn't seem as promising from your perspective.

Cool. I don't have a timeline for when I will be available to do this but I would be happy to try and set something up or at least get started with it. The `Onigmo` tests look really promising and I think if we have reusable machinery it would make sense to add some of the others.
